File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsValidBST(TreeNode root) {
        if(root.left == null && root.right == null) return true;
        return Validate(root,int.MinValue,int.MaxValue, false,false);
    }

    private bool Validate(TreeNode node, int low, int high, bool lowBoundExists, bool highBoundExists) {
        if (node == null) return true;

        // Check if the node's value violates the low or high boundaries
        if ((lowBoundExists && node.val <= low) || 
            (highBoundExists && node.val >= high)) {
            return false;
        }

        // Recursively validate the left and right subtrees
        return Validate(node.left, low, node.val, lowBoundExists, true) && 
               Validate(node.right, node.val, high, true, highBoundExists);
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:


        return self.dfs(root, 99999999999999999999,-9999999999999999999999)


    def dfs(self, node, maxi, mini):

        if not node:
            return True
        
        if node.val <= mini or node.val >= maxi:
            return False
        
        return self.dfs(node.left, node.val, mini) and self.dfs(node.right, maxi, node.val)


        

            

        

            
        